botcore/utils/cooldown.py

```diff
@@ -0,0 +1,184 @@
+"""Helpers for setting a cooldown on commands."""
+
+from __future__ import annotations
+
+import asyncio
+import math
+import random
+import time
+import typing
+from collections.abc import Awaitable, Hashable
+from contextlib import suppress
+from dataclasses import dataclass
+from typing import Callable  # sphinx-autodoc-typehints breaks with collections.abc.Callable
+
+import discord
+from discord.ext.commands import CommandError, Context
+
+from botcore.utils import scheduling
+from botcore.utils.function import command_wraps
+
+__all__ = ["CommandOnCooldown", "block_duplicate_invocations", "P", "R"]
+
+_ArgsTuple = tuple[object]
+
+if typing.TYPE_CHECKING:
+    from botcore import BotBase
+    import typing_extensions
+    P = typing_extensions.ParamSpec("P")
+    P.__constraints__ = ()
+else:
+    P = typing.TypeVar("P")
+    """The command's signature."""
+
+R = typing.TypeVar("R")
+"""The command's return value."""
+
+
+class CommandOnCooldown(CommandError, typing.Generic[P, R]):
+    """Raised when a command is invoked while on cooldown."""
+
+    def __init__(
+        self,
+        message: str | None,
+        function: Callable[P, Awaitable[R]],
+        *args: object,
+        **kwargs: object,
+    ):
+        super().__init__(message, function, args, kwargs)
+        self._function = function
+        self._args = args
+        self._kwargs = kwargs
+
+    async def call_without_cooldown(self) -> R:
+        """
+        Run the command this cooldown blocked.
+
+        Returns:
+            The command's return value.
+        """
+        return await self._function(*self._args, **self._kwargs)
+
+
+@dataclass
+class _CooldownItem:
+    call_arguments: _ArgsTuple
+    timeout_timestamp: float
+
+
+class _CommandCooldownManager:
+    """
+    Manage invocation cooldowns for a command through the arguments the command is called with.
+
+    A cooldown is set through `set_cooldown` for a channel with the given `call_arguments`,
+    if `is_on_cooldown` is checked within `cooldown_duration` seconds
+    of the call to `set_cooldown` with the same arguments, True is returned.
+    """
+
+    def __init__(self, *, cooldown_duration: float):
+        self._cooldowns = dict[tuple[Hashable, _ArgsTuple], float]()
+        self._cooldowns_non_hashable = dict[Hashable, list[_CooldownItem]]()
+        self._cooldown_duration = cooldown_duration
+        self.cleanup_task = scheduling.create_task(
+            self._periodical_cleanup(random.uniform(0, 10)),
+            name="CooldownManager cleanup",
+        )
+
+    def set_cooldown(self, channel: Hashable, call_arguments: _ArgsTuple) -> None:
+        """Set `call_arguments` arguments on cooldown in `channel`."""
+        timeout_timestamp = time.monotonic() + self._cooldown_duration
+
+        try:
+            self._cooldowns[(channel, call_arguments)] = timeout_timestamp
+        except TypeError:
```

Comment:
In what case is a `TypeError` raised?

Reply:
Maybe I'm dumb but I found that confusing despite the name of the dict. I think it would help to have a comment pointing out that call arguments aren't always going to be hashable (maybe better to put that in the dict declaration in `__init__).